src/test/cypress/support/requests/CourseManagementRequests.ts

```diff
@@ -59,17 +59,31 @@ export class CourseManagementRequests {
      * @param packageName the package name of the programming exercise
      * @param releaseDate when the programming exercise should be available (default is now)
      * @param dueDate when the programming exercise should be due (default is now + 1 day)
+     * @param group the exercise group object if the exercise is in an exam
+     * @param buildAndTestDate the date for automated building and testing after exam completion
      * @returns <Chainable> request response
      */
-    createProgrammingExercise(course: any, title: string, programmingShortName: string, packageName: string, releaseDate = new Date(), dueDate = new Date(Date.now() + oneDay)) {
-        const programmingTemplate = programmingExerciseTemplate;
+    createProgrammingExercise(
+        title: string,
+        programmingShortName: string,
+        packageName: string,
+        course?: any,
+        group?: any,
+        buildAndTestDate?: day.Dayjs,
+        releaseDate = new Date(),
+        dueDate = new Date(Date.now() + oneDay),
+    ) {
```

Comment:
You can replace the two optional parameters with two required parameters. Unfortunately I can't give a proper code suggestion because github doesn't allow suggestions on unchanged code lines. I pasted the suggestion below, but you can't directly commit it. This suggestions would work in combination with the suggestion below for the `getCourseOrExamExercise` method. Also the `POST` string can be replaced by referencing the string in the constants file
```suggestion
    /**
     * Creates a course with the specified title and short name.
     * @param course the response object from a previous call to createCourse
     * @param title the title of the programming exercise
     * @param programmingShortName the short name of the programming exercise
     * @param packageName the package name of the programming exercise
     * @param releaseDate when the programming exercise should be available (default is now)
     * @param dueDate when the programming exercise should be due (default is now + 1 day)
     * @param body the exercise group object or the course object
     * @param isExamExercise if the programming exercise is part of an exam or not
     * @param buildAndTestDate the date for automated building and testing after exam completion
     * @returns <Chainable> request response
     */
    createProgrammingExercise(
        title: string,
        programmingShortName: string,
        packageName: string,
        body: any,
        isExamExercise: boolean,
        buildAndTestDate?: day.Dayjs,
        releaseDate = new Date(),
        dueDate = new Date(Date.now() + oneDay),
    ) {
        const programmingTemplate = this.getCourseOrExamExercise(programmingExerciseTemplate, body, isExamExercise);
        programmingTemplate.title = title;
        programmingTemplate.shortName = programmingShortName;
        programmingTemplate.packageName = packageName;
        if (!isExamExercise) {
            programmingTemplate.releaseDate = releaseDate.toISOString();
            programmingTemplate.dueDate = dueDate.toISOString();
        } else {
            programmingTemplate.allowComplaintsForAutomaticAssessments = true;
            programmingTemplate.buildAndTestStudentSubmissionsAfterDueDate = buildAndTestDate;
        }
        return cy.request({
            url: PROGRAMMING_EXERCISE_BASE + 'setup',
            method: POST,
            body: programmingTemplate,
        });
    }
```

Reply:
Maybe we can even determine which type of object `body` is at runtime. For example an exam object would have the `randomizeExerciseOrder` field, but the course not. Depending on this we could also switch the logic. Then we can get rid of the `isExamExercise` field and only need the `body` field

Reply:
i think i refactored this to your satisfaction